engine/src/main/java/org/terasology/rendering/dag/RenderTaskListGenerator.java

```diff
@@ -19,117 +19,212 @@
-                            // non-redundant state change of the same subType but different value, becomes new map entry
-                            taskList.add(stateChange.generateTask());
-                            persistentStateChanges.put(stateChange.getClass(), stateChange);
-
-                        } // else {
-                        // the non-redundant state change being examined is not a default one and is identical
-                        // in value to the one stored in the persistentStateChange map: it is redundant after all
-                        // and we ignore it
-                        // }
+                // task executing the node.process() method
+                taskList.add(node.generateTask());
+
+                // generating tasks for the desired state resets
+                nextEnabledNode = findNextEnabledNode(orderedNodes, currentIndex + 1);
+                if (nextEnabledNode != null) {
+
+                    // if there is one enabled node after this one we must check its desired state changes,
+                    // to make sure we don't reset to default something it would set again.
+                    // For example: no point binding the display (the default FBO) if the next enabled node
+                    // will bind another FBO.
+                    for (StateChange currentStateReset : node.getDesiredStateResets()) {
+                        if (sameClassStateChangeNotFoundInThe(nextEnabledNode, currentStateReset)) {
+                            taskList.add(currentStateReset.generateTask());
+                            persistentStateChanges.remove(currentStateReset.getClass());
+                        }
                     }
 
+                } else {
+                    // there are no enabled nodes after this one: we must generate all the tasks necessary
+                    // to reset the effects of any persisting state change.
+                    for (StateChange currentStateReset : node.getDesiredStateResets()) {
+                        taskList.add(currentStateReset.generateTask());
+                        persistentStateChanges.remove(currentStateReset.getClass());
+                    }
                 }
+            }
+
+            currentIndex++;
+        }
 
-                intranodesStateChanges.clear();
-                taskList.add(((Node) object).generateTask());
+        long endTimeInNanoSeconds = System.nanoTime();
+
+        if (logger.isInfoEnabled()) {
+            logger.info("===== INTERMEDIATE RENDERER LIST =========================");
+            logIntermediateRendererListForDebugging(orderedNodes);
+            logger.info("===== RENDERER TASK LIST =================================");
+            logList(taskList);
+            logger.info("----------------------------------------------------------");
+            logger.info(String.format("Task list generated in %.3f ms", endTimeInNanoSeconds - startTimeInNanoSeconds / 1000000f));
```

Comment:
As this line is written, it's performing the calculation:

```
endTimeInNanoSeconds - (startTimeInNanoSeconds / 1000000f)
```

Which, on my machine, is producing something like:

```
23:01:14.514 [main] INFO  o.t.r.dag.RenderTaskListGenerator - Task list generated in 505639151337472.000 ms
```

A simple fix is to change this line to:

```
logger.info(String.format("Task list generated in %.3f ms", (endTimeInNanoSeconds - startTimeInNanoSeconds) / 1000000f));
```

which appears to solve the issue:

```
23:13:11.863 [main] INFO  o.t.r.dag.RenderTaskListGenerator - Task list generated in 3.043 ms
```